experimental/displaynames/src/displaynames.rs

```diff
@@ -404,90 +405,216 @@ impl LocaleDisplayNamesFormatter {
     }
 
     /// Returns the display name of a locale.
+    /// This implementation is based on the algorithm described in
+    /// https://www.unicode.org/reports/tr35/tr35-general.html#locale_display_name_algorithm
+    ///
     // TODO: Make this return a writeable instead of using alloc
     pub fn of<'a, 'b: 'a, 'c: 'a>(&'b self, locale: &'c Locale) -> Cow<'a, str> {
-        // https://www.unicode.org/reports/tr35/tr35-general.html#Display_Name_Elements
+        let longest_matching_identifier = self.find_longest_matching_subtag(&locale);
+
+        // Step - 1: Construct a locale display name string (LDN).
+        // Find the displayname for the longest_matching_subtag which was derived above.
+        let ldn = self.get_locale_display_name(&locale, &longest_matching_identifier);
+
+        // Step - 2: Construct a vector of longest qualifying substrings (LQS).
+        // Find the displayname for the remaining locale if exists.
+        let lqs = self.get_longest_qualifying_substrings(&locale, &longest_matching_identifier);
+
+        // Step - 3: Return the displayname based on the size of LQS.
+        let mut result = Cow::Borrowed(ldn);
+        if lqs.len() > 0 {
+            let mut output = String::with_capacity(
+                result.len() + " (".len() + lqs.iter().map(|s| ", ".len() + s.len()).sum::<usize>()
+                    - ", ".len()
+                    + ")".len(),
+            );
+            output.push_str(&result);
+            output.push_str(" (");
+            output.push_str(lqs[0]);
+            for lqs in &lqs[1..] {
+                output.push_str(", ");
+                output.push_str(lqs);
+            }
+            output.push_str(")");
+            result = Cow::Owned(output);
+        }
+        result
+    }
 
-        // TODO: This binary search needs to return the longest matching found prefix
-        // instead of just perfect matches
-        if let Some(displayname) = match self.options.style {
-            Some(Style::Short) => self
+    /// For a given locale and the data, find the longest prefix of the string that exists as a key in the CLDR locale data.
+    pub fn find_longest_matching_subtag(&self, locale: &Locale) -> LanguageIdentifier {
```

Comment:
Observation: this only ever returns one of the following:

- language-script
- language-region
- language

But does the spec allow us to return language-script-region, etc? Consider this in #3913